doc/man-source/ircat.1
.TH IRCAT "1" "Last change: Jan 2017" "ircat @version@" "User Commands"
.SH NAME
ircat - print strings when pressing buttons
.SH SYNOPSIS
.B ircat
[\fIoptions\fR] <\fIprog\fR>
.SH DESCRIPTION
This program prints config strings to standard output. It can be used
to provide remote control input to scripts and to debug your .lircrc
file.

The argument to the program is the program name, as it
appears in the prog entries in .lircrc.
.TP
\fB\-h\fR \fB\-\-help\fR
Display usage summary.
.TP
\fB\-v\fR \fB\-\-version\fR
Display version.
.TP
\fB\-c\fR \fB\-\-config=\fR<\fIfile\fR>
Set config file.
.SH EXAMPLES
If .lircrc contains:

.nf
begin
        prog = myprog
        button = KEY_CHANNELUP
        config = next_file
end
.fi

then

.nf
$ ircat myprog
.fi

will print "next_file" (followed by newline) every time the
button bound to KEY_CHANNELUP is pressed.
.SH ENVIRONMENT
.TP 4
.B LIRC_SOCKET_PATH
The \fBlircd(8)\fR output socket used to retrieve data.
Defaults to @varrundir@/lirc/lircd

.SH "SEE ALSO"
.P
The documentation for
.B lirc
is maintained as html pages. They are located under html/ in the
documentation directory.


NEWS
0.9.4d 23/01/1e7
 * The debian packaging means that freebsd and hurd builds are
   regularly tested, improving build stability.
 * A multitude of build patches, mostly after building on Debian:
    + Cross-compilation FTBS fixed (#233)
    + Fix configuration of configure (#234)
    + Fix library specification in lirc.pc (#236)
    + use HAVE_UINPUT=1 to force building uinput code (#238).
    + lib: Only use local headers in lirc_private.h (#237).
    + Configure libudev (#242)
    + Various kfreebsd fixes
    + Include config.h in installed headers (#250).
    + Fix _XOPEN_SOURCE related FTBS on some platforms (#248)
 * lib: Fix some lircrcd-related bugs (#254)
 * systemd: Fix wrong path to irexec (#239)
 * systemd: Add missing lircd-setup.service to distribution (#243)
 * lircd: Fix --listen option parsing [#249]
 * lircd: Fix segfault in SET_INPUTLOG without arguments (#252)
 * lircd: Fix DRV_OPTION error message handling (#256).
 * ircat: Fix option parsing bug + empty conditions (#251)
 * xmode2: Fix handling of --driver option.

0.9.4c 22/10/16
 * New curl_poll files adds a new MIT license. Lirc now carries
   GPL, BSD and MIT licenses.
 * Several patches to make lirc work again on macos (#226, #227).
 * New tool irtext2udp generates udp plugin data (#197).
 * Fix FTBS errors on linux kernel >= 4.8.0 (#232).
 * plugins: girs: Add udp/receive support (#201).
 * mode2: Fix broken --raw option (#223).
 * lirc-make-devinput: Fix that two remotes has the same name (#222).
 * lircd: Warn for duplicate remote definitions (#222).
 * lircd: Retry temporary unavailable write sockets (#221).
 * contrib: Fix usb devices' acl permissions (rhbz #1364744).
 * lirc-setup: Fix crasher bugfix (Fedora abort trace).
 * contrib: disable udev /dev/uinput rule requiring lirc user.
 * plugins: audio_alsa: Fix byte truncating in 16-bit data (#218):
 * plugins: irtoy: improve error handling (#220).
 * Minor documentation fixes.

0.9.4b 9/8/16
  * lircd-uinput: A multitude of fixes (#213, #211, #177). Adds new options.
  * lirc_client:  Fix "Can't redirect irsend output" (#207).
  * lirc_client: read_string() not terminating buffer contents (#216)
  * Remove a single python2 dependency [4e3c97c].
  * irrecord: Fix under-sized buffer causing crash (#208).
  * systemd: Add hardening options comments to services (#204).
  * lircmd: Options, udev examples and manpage update.A
  * lib: Add missing Makefile.am dep (#210).
  * doc: Manpages and configuration guide fixes.
  * build: Don't define docdir - autoconf defines docdir (#215).
  * build: support runstatedir configure option (#212).

0.9.4a 22/6/16
  * Fix syntax errors in systemd unit files (rhbz: 1350750).
  * irrecord: Enhance "Don't use devinput" message (#180).
  * plugins/uirt2_raw: Reset device to UIR mode on deinit.
  * plugins/girs.c: Enable reception after sending.
  * lirc.org: Updates and bugfixes
  * lircd, liblirc_client: Fix freeaddrinfo handling (#195).
  * Don't build commandir unless we have usb.h (#191).
  * lirc_options.conf: debug => loglevel (#199).
  * lib: use proper linking method to avoid parallel build issue
  * tools: make_rel_symlink.py can also use python2
  * plugins/devinput: Make the list-devices support available.
  * configure.ac: Bugfix (BSD setting reported wrong).

0.9.4 22/5/16
  * Issues: https://sf.net/p/lirc/tickets/milestone/0.9.4
  * Digest of 0.9.4pre1, 0.9.4pre2

  * The caraca driver has been retired (still available in attic/).
  * The iguanair driver has been moved out of tree and is now
    maintained by Iguanaworks, Inc. See drivers/iguanair/README.
  * The irman driver has been moved out of tree. From 0.5.1 the
    libirman package contains the driver. See drivers/irman/README
  * The yard2 driver has been moved out of tree and is now maintained
    by the yard2 project. See driver/yard2/README.
  * The lirc.4 manpage (/dev/lirc* interface) is upstreamed to the
    linux kernel manpages project (#164). Old page lirc.4l still
    lingers in attic/ but is not installed.
  * Upstream selinux updated to match current lircd requirements.
    See: https://github.com/fedora-selinux/selinux-policy/pull/41

  * New upstream debian packaging with which Debian users can create
    .deb packages. Ubuntu packages are also available in PPAs. See
    README.md in the debian branch (#137).
  * New driver girs for the GIRS protocol, gives LIRC Arduino support.
    See new info in driver documentation.
  * New transmit-only FTDI driver with enhanced accuracy and hardware
    support. See ftdix driver documentation.
  * Support for non-modulated signals added (#132).
  * As shipped, lircd now uses the devinput driver and the new 'auto'
    device, should work for all remotes supported by kernel (#106).
  * Enhanced lirc-setup tool (#123, partly):
    - Add drvctl device enumeration support.
    - Integrated mode2 and irw tests of current configuration.
    - Documentation links.
    - More consistent data based on new database.
    - Multiple bugfixes.
  * Several patches to make mode2 work with all drivers(#123, partly)
  * Configuration database overhaul (#123, partly)
    - Database lives as yaml files in configs/.
    - Data is collected from drivers, remotes and static files.
    - Better ways for drivers to enumerate devices.
    - There is an unstable python database module used by
      all tools.
  * Several drivers including default and devinput has enhanced
    enumeration support, usable with mode2 -l and lirc-setup.
  * devinput.lircd.conf is generated at build time to match host kernel,
    new script lirc-make-devinput shipped to re-generate it (#153)
  * A new script lircd-setup runs as a one-shot service before lircd,
    performing setup commands as root (#121).
  * Several new examples in contrib/, notably lircrc templates for
    many popular LIRC applications and improved udev rules.
  * The default driver does not require root perms if the protocol
    is correct (#133).

tools/ircat.cpp
/****************************************************************************
** ircat.c *****************************************************************
****************************************************************************
*
* ircat - prints config strings to standard output, can be used to
* provide remote control input to scripts
*
* The first agrument to the program is the program name, as it
* appears in the prog entries in .lircrc.
*
* For example if .lircrc contains:
*
* begin
*       prog = myprog
*       button = tv_p+
*       config = next_file
* end
*
* then
*
* $ ircat myprog
*
* will print "next_file" (followed by newline) every time the
* button tv_p+ is pressed.
*
*
* Copyright (C) 2002 Bjorn Bringert <bjorn@bringert.net>
*
* Based on irexec.c
*
*/

#ifdef HAVE_CONFIG_H
# include <config.h>
#endif

#include <errno.h>
#include <unistd.h>
#include <stdarg.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <getopt.h>
#include "lirc_client.h"

#define PROG_NAME "ircat"
#define PROG_VERSION (PROG_NAME VERSION)

void print_usage(char* prog_name)
{
	printf("Usage: %s [options] <prog>\n", prog_name);
	printf("\t -h --help\t\tdisplay usage summary\n");
	printf("\t -v --version\t\tdisplay version\n");
	printf("\t -c --config=<file>\tset config file\n");
}

int main(int argc, char* argv[])
{
	struct lirc_config* config;
	char* config_file = NULL;
	int r;

	while (1) {
		int c;
		static struct option long_options[] = {
			{ "config",  required_argument, NULL, 'c' },
			{ "help",    no_argument,	NULL, 'h' },
			{ "version", no_argument,	NULL, 'v' },
			{ 0,	     0,			0,    0	  }
		};
		c = getopt_long(argc, argv, "c:hv", long_options, NULL);
		if (c == -1)
			break;
		switch (c) {
		case 'c':
			config_file = optarg;
			break;
		case 'h':
			print_usage(argv[0]);
			return EXIT_SUCCESS;
		case 'v':
			printf("%s\n", PROG_VERSION);
			return EXIT_SUCCESS;
		default:
			print_usage(argv[0]);
			return EXIT_FAILURE;
		}
	}

	if (optind != argc - 1) {
		print_usage(argv[0]);
		return EXIT_FAILURE;
	}

	if (lirc_init(argv[argc - 1], 1) == -1)
		exit(EXIT_FAILURE);

	r = lirc_readconfig(config_file, &config, NULL);
	if (r == 0) {
		char* code;
		char* c;
		int r;

		while (lirc_nextcode(&code) == 0) {
			if (code == NULL || !*code)
				continue;
			while ((r = lirc_code2char(config, code, &c)) == 0) {
				if (c == NULL || !*c)
					break;
				printf("%s\n", c);
				fflush(stdout);
			}
			free(code);
			if (r == -1)
				break;
		}
		lirc_freeconfig(config);
	}

	lirc_deinit();
	exit(EXIT_SUCCESS);
}
